server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java

```diff
@@ -123,6 +135,70 @@ protected SearchPhase getNextPhase(SearchPhaseResults<CanMatchResponse> results,
         return new GroupShardsIterator<>(sortShards(shardsIts, results.minAndMaxes, fieldSort.order()));
     }
 
+    /**
+     * Iterate over the shard iterator and look for overlapping rollup indices. If rollup indices exist,
+     * the method includes the shards from the optimal rollup index that can match the query and ignores
+     * the shards from other overlapping indices.
+     *
+     * @return A {@link FixedBitSet} with the shards of the optimal indices.
+     */
+    private FixedBitSet chooseOptimalRollupShards(GroupShardsIterator<SearchShardIterator> shardsIts, FixedBitSet possibleMatches,
+                                                  ClusterState clusterState, SearchSourceBuilder source) {
+        // Map with key the index that will be replaced by an optimal rollup index. The optimal index
+        // will be its replacement. In the end all indices contains in the keys will be ignored.
+        Map<String, Tuple<String, RollupIndexMetadata>> indexReplacements = new HashMap<>(shardsIts.size());
+
+        int i = 0;
+        for (SearchShardIterator iter : shardsIts) {
+            if (possibleMatches.get(i++)) {
+                String indexName = iter.shardId().getIndexName();
+                Map<String, String> idxMetadata = clusterState.getMetadata().index(indexName).getCustomData(RollupIndexMetadata.TYPE);
+                IndexAbstraction originalIndex = clusterState.getMetadata().getIndicesLookup().get(indexName);
+
+                // If index is not a member of a datastream it will not be replaced
```

Comment:
I think it could be useful to mention that the index is also not associated to a rollup group, since that is when `idxMetadata` would be `null`.